src/irr/asset/CGraphicsPipelineLoaderMTL.cpp

```diff
@@ -1,10 +1,393 @@
+        N = normalize(N - surfGrad);
+    }
+    vec3 L = normalize(-ViewPos);
+    vec3 R = normalize(-reflect(L,N));
+    float NdotL = max(dot(N,L), 0.0);
+    float VdotR = max(dot(L,R), 0.0);
+
+    vec3 Ke = texture(map_Ke, UV).rgb;//force visibility (as in: visible as resource used by the shader) of emissive map (needed temporarily because of bindings reordering bug)
+
+    vec3 Kd;
+    if ((PC.extra&(map_Kd_MASK)) == (map_Kd_MASK))
+        Kd = texture(map_Kd, UV).rgb;
+    else
+        Kd = PC.Kd;
+    float d = 1.0;
+    if ((PC.extra&(map_d_MASK)) == (map_d_MASK))
+        d = texture(map_d, UV).r;
+
+    vec3 color;
+    if ((PC.extra&ILLUM_MODEL_MASK) > 0)
+    {
+        vec3 Ka;
+        vec3 Ks;
+        float Ns;
+        if ((PC.extra&(map_Ka_MASK)) == (map_Ka_MASK))
+            Ka = texture(map_Ka, UV).rgb;
+        else
+            Ka = PC.Ka;
+        if ((PC.extra&(map_Ks_MASK)) == (map_Ks_MASK))
+            Ks = texture(map_Ks, UV).rgb;
+        else
+            Ks = PC.Ks;
+        if ((PC.extra&(map_Ns_MASK)) == (map_Ns_MASK))
+            Ns = texture(map_Ns, UV).x;
+        else
+            Ns = PC.Ns;
+
+        color = Ka*Ia + Kd*NdotL;
+        switch (PC.extra&ILLUM_MODEL_MASK)
+        {
+        case 2:
+        case 3://2 + reflection map
+        case 4://3 with transparency (glass)
+        case 6:
+        case 8://reflection map
+        case 9://reflection map
+            color += Ks*pow(VdotR, Ns);
+            break;
+        case 5:
+        case 7:
+            color += Ks*pow(VdotR, Ns)*Fresnel_dielectric(PC.Ni, NdotL);
+            break;
+        default:
+            break;
+        }
+    }
+    else color = Kd;
+
+    OutColor = vec4(color, d);
+}
+#endif //_IRR_FRAG_MAIN_DEFINED_
+)";
+}
+
+
 using namespace irr;
 using namespace asset;
 
+static void insertShaderIntoCache(core::smart_refctd_ptr<ICPUSpecializedShader>& asset, const char* path, IAssetManager* _assetMgr)
+{
+    asset::SAssetBundle bundle({ asset });
+    _assetMgr->changeAssetKey(bundle, path);
+    _assetMgr->insertAssetIntoCache(bundle);
+}
+
+_IRR_STATIC_INLINE_CONSTEXPR const char* VERT_SHADER_NO_UV_CACHE_KEY = "irr/builtin/shader/loaders/mtl/vertex_no_uv";
+_IRR_STATIC_INLINE_CONSTEXPR const char* VERT_SHADER_UV_CACHE_KEY = "irr/builtin/shader/loaders/mtl/vertex_uv";
+_IRR_STATIC_INLINE_CONSTEXPR const char* FRAG_SHADER_NO_UV_CACHE_KEY = "irr/builtin/shader/loaders/mtl/fragment_no_uv";
+_IRR_STATIC_INLINE_CONSTEXPR const char* FRAG_SHADER_UV_CACHE_KEY = "irr/builtin/shader/loaders/mtl/fragment_uv";
+
 CGraphicsPipelineLoaderMTL::CGraphicsPipelineLoaderMTL(IAssetManager* _am) : m_assetMgr{_am}
 {
+    //create vertex shaders and insert them into cache
+    {
+        std::string vs_nouv_source = VERT_SHADER_UV;
+        vs_nouv_source.insert(vs_nouv_source.find('\n') + 1ull,
+            R"(
+#define _IRR_VERT_MAIN_DEFINED_
+)"
+);
+        vs_nouv_source.insert(vs_nouv_source.size(), VERT_SHADER_NO_UV_MAIN);
```

Comment:
you also need to override the vertex shader ouputs, you can't have a dangling UV out.

Reply:
done?